### PR TITLE
Introduce `MockableDatabase` for wallet unit tests

### PR DESCRIPTION
**Pull request by achow101**
For the wallet's unit tests, we currently use either `DummyDatabase` or memory-only versions of either BDB or SQLite. The tests that use `DummyDatabase` just need a `WalletDatabase` so that the `CWallet` can be constructed, while the tests using the memory-only databases just need a backing data store. There is also a `FailDatabase` that is similar to `DummyDatabase` except it fails be default or can have a configured return value. Having all of these different database types can make it difficult to write tests, particularly tests that work when either BDB or SQLite is disabled.

This PR unifies all of these different unit test database classes into a single `MockableDatabase`. Like `DummyDatabase`, most functions do nothing and just return true. Like `FailDatabase`, the return value of some functions can be configured on the fly to test various failure cases. Like the memory-only databases, records can actually be written to the `MockableDatabase` and be retrieved later, but all of this is still held in memory. Using `MockableDatabase` completely removes the need for having BDB or SQLite backed wallets in the unit tests for the tests that are not actually testing specific database behaviors.

Because `MockableDatabase`s can be created by each unit test, we can also control what records are stored in the database. Records can be added and removed externally from the typical database modification functions. This will give us greater ability to test failure conditions, particularly those involving corrupted records.

Possible alternative to #26644

### CODE REVIEW

File: src/wallet/salvage.h [hunk 1]
```diff
@@ -8,12 +8,60 @@
...
+
+    std::unique_ptr<DatabaseCursor> GetNewCursor() override { return std::make_unique<DummyCursor>(); }
+    bool TxnBegin() override { return true; }
+    bool TxnCommit() override { return true; }
+    bool TxnAbort() override { return true; }
+};
+
+/** A dummy WalletDatabase that does nothing and never fails. Only used by salvage.
+ **/
+class DummyDatabase : public WalletDatabase
```
Comment (TheCharlatan) [inline src/wallet/salvage.h:45]: Couldn't this class definition be moved out of the header?

File: src/wallet/salvage.cpp [hunk 2]
```diff
@@ -167,4 +167,10 @@ bool RecoverDatabaseFile(const ArgsManager& args, const fs::path& file_path, bil
 
     return fSuccess;
 }
+
+/** Return object for accessing dummy database with no read/write capabilities. */
+std::unique_ptr<WalletDatabase> CreateDummyWalletDatabase()
+{
+    return std::make_unique<DummyDatabase>();
```
Comment (TheCharlatan) [inline src/wallet/salvage.cpp:174]: Nit: Couldn't this just be used directly at its single call site?

**Review by sedited — COMMENTED** [review]
Approach ACK

Just some minor comments; I like this change. I did not check yet if there are other tests that might be simpler with the new mockable class.

File: src/wallet/salvage.h (hunk 1 above)
Comment (achow101) [reply src/wallet/salvage.h:45]: Done

File: src/wallet/salvage.cpp (hunk 2 above)
Comment (achow101) [reply src/wallet/salvage.cpp:174]: Done

File: src/wallet/salvage.h (hunk 1 above)
Comment (TheCharlatan) [reply src/wallet/salvage.h]: ACK, this is resolved.

File: src/wallet/salvage.cpp (hunk 2 above)
Comment (TheCharlatan) [reply src/wallet/salvage.cpp]: ACK, this is resolved.

**Review by furszy — COMMENTED** [review]
Concept ACK, great idea to make db independent tests.

**Review by furszy — COMMENTED** [review]
Code review ACK a9c58a93 with something that we should correct here and improve in a follow-up after #24914:

Need to restore `m_records` to previous state after each scenario in `wallet_load_ckey`.

The reason is that we are currently not checking for specific error messages, we only check that `LoadWallet` returns a `DBErrors::CORRUPT` error. So, follow-up test cases could just be carrying previous corruption errors and not the one that we are trying to test.

Note: This is currently fine because errors are tested in-order (checksum error is tested prior to pubkey corruption error) but could not be the case in the future.

After #24914, we could improve it to check against the specific error message.

**Review by sedited — APPROVED** [review]
Code review ACK [a9c58a9](https://github.com/bitcoin/bitcoin/pull/26715/commits/a9c58a93bc4929724ff539bb5be457e8b2eb7dc0)

File: src/wallet/test/util.cpp [hunk 3]
```diff
@@ -7,6 +7,7 @@
 #include <chain.h>
 #include <key.h>
 #include <key_io.h>
+#include <streams.h>
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/wallet/test/util.cpp:10]: why not? `DataStream` is defined there.

[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/wallet/test/util.cpp:10]: Yeah totally, not sure what I was thinking here.
Please resolve this.

File: src/wallet/test/wallet_tests.cpp [hunk 4]
```diff
@@ -1008,7 +960,7 @@ BOOST_FIXTURE_TEST_CASE(wallet_sync_tx_invalid_state_test, TestingSetup)
     // 1) Make db always fail
     // 2) Try to add a transaction that spends the previously created transaction and
     //    verify that we are not moving forward if the wallet cannot store it
-    static_cast<FailDatabase&>(wallet.GetDatabase()).m_pass = false;
+    static_cast<MockableDatabase&>(wallet.GetDatabase()).m_pass = false;
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/test/wallet_tests.cpp]: Done as suggested

File: src/wallet/salvage.h [hunk 5]
```diff
@@ -8,6 +8,7 @@
 
 #include <fs.h>
 #include <streams.h>
+#include <wallet/db.h>
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/salvage.h]: Removed

File: src/wallet/test/util.cpp [hunk 6]
```diff
@@ -79,4 +61,92 @@ CTxDestination getNewDestination(CWallet& w, OutputType output_type)
...
+    auto it = m_records.begin();
+    while (it != m_records.end()) {
+        auto& key = it->first;
+        if (key.size() < prefix.size() || std::search(key.begin(), key.end(), prefix.begin(), prefix.end()) == key.begin()) {
+            it++;
+            continue;
+        }
+        it = m_records.erase(it);
+    }
+}
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/test/util.cpp]: Oops, fixed

File: src/wallet/test/util.cpp [hunk 7]
```diff
@@ -79,4 +80,93 @@ CTxDestination getNewDestination(CWallet& w, OutputType output_type)
...
+    if (!m_pass) {
+        return false;
+    }
+    auto it = m_records.begin();
+    while (it != m_records.end()) {
+        auto& key = it->first;
+        if (key.size() < prefix.size() || std::search(key.begin(), key.end(), prefix.begin(), prefix.end()) == key.begin()) {
+            it++;
+            continue;
+        }
```
Comment (furszy) [inline src/wallet/test/util.cpp:138]: if found, `std::search` returns the Iterator to the beginning of the first occurrence of the sequence, so this should be `!= key.begin` or `== key.end()`.

Crafted a quick test to check it https://github.com/furszy/bitcoin-core/commit/f8a7c0670f95b317c46b01855c4c0f0db7ba7714 (no need to add it, I was just double checking it)

File: src/bench/wallet_loading.cpp [hunk 8]
```diff
@@ -71,7 +71,7 @@ static void WalletLoading(benchmark::Bench& bench, bool legacy_wallet)
         options.create_flags = WALLET_FLAG_DESCRIPTORS;
         options.require_format = DatabaseFormat::SQLITE;
     }
-    auto database = CreateMockWalletDatabase(options);
+    auto database = CreateMockableWalletDatabase();
```
Comment (furszy) [inline src/bench/wallet_loading.cpp:74]: What about the db `require_format` that is above? Is that used anywhere?

File: src/wallet/test/util.cpp (hunk 7 above)
Comment (achow101) [reply src/wallet/test/util.cpp]: Fixed to `!=`

File: src/bench/wallet_loading.cpp (hunk 8 above)
Comment (achow101) [reply src/bench/wallet_loading.cpp]: Added a commit dropping those.

File: src/bench/wallet_loading.cpp [hunk 9]
```diff
@@ -65,11 +64,8 @@ static void WalletLoading(benchmark::Bench& bench, bool legacy_wallet)
...
     DatabaseOptions options;
-    if (legacy_wallet) {
-        options.require_format = DatabaseFormat::BERKELEY;
-    } else {
+    if (!legacy_wallet) {
         options.create_flags = WALLET_FLAG_DESCRIPTORS;
-        options.require_format = DatabaseFormat::SQLITE;
     }
     auto database = CreateMockableWalletDatabase();
     auto wallet = BenchLoadWallet(std::move(database), context, options);
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/bench/wallet_loading.cpp]: Done

**Review by furszy — APPROVED** [review]
ACK 33e2b82

**Review by sedited — APPROVED** [review]
ACK 33e2b82a4fc990253ff77655f437c7aed336bc55

**Comment by kouloumos** [comment]
While reviewing https://github.com/bitcoin/bitcoin/pull/27666, I observed that this PR changed the performance of the `WalletLoadingLegacy` benchmark:

PR branch (33e2b82a4fc990253ff77655f437c7aed336bc55)
|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|       63,589,392.00 |               15.73 |    0.1% |  265,707,729.00 |  125,674,284.00 |  2.114 |  32,837,198.00 |    1.0% |      0.32 | `WalletLoadingLegacy`

master at the time (539452242e895f2dcd719d41f447a48896d0e4b2)
|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|      272,970,312.00 |                3.66 |    0.1% |  617,971,997.00 |  417,477,301.00 |  1.480 | 103,288,956.00 |    0.4% |      1.39 | `WalletLoadingLegacy``

This is probably not an issue, but it seems worth noting that a change of the underlying mock implementation had such an impact on the performance of legacy wallet loading. I am thinking that this might be the case of a not so effective benchmark, but I'll need to investigate further as I haven't look into the actual changes of this PR yet.

**Comment by achow101** [comment]
> This is probably not an issue, but it seems worth noting that a change of the underlying mock implementation had such an impact on the performance of legacy wallet loading.

It changed the underlying db used for this benchmark from the BDB in-memory database to the MockableDatabase. However I think the goal of this benchmark is to measure `LoadWallet` specifically and not necessarily the database engine, so I don't think it particularly matters.

**Pull request merged**